lib/phoenix/router.ex

```diff
@@ -357,6 +358,10 @@ defmodule Phoenix.Router do
                  unquote(exprs.host)) do
         unquote(exprs.dispatch)
       end
+      def __match_params__(_conn, unquote(exprs.verb_match), unquote(exprs.path),
```

Comment:
I don't like that we are making the router larger by adding all of those new clauses. But it seems we can't use the dispatch one because it invokes the pipe_through stuff?

About the name, one other option is to make this API public. The reason I like `__routes__` is because if one day we want to move the generators out of Phoenix it won't be a problem.